src/Microsoft.NetCore.Analyzers/Core/Security/DoNotCallDangerousMethodsInDeserialization.cs

```diff
@@ -0,0 +1,243 @@
+
+                    if (onDeserializedAttributeTypeSymbol != null)
+                    {
+                        attributeTypeSymbolsBuilder.Add(onDeserializedAttributeTypeSymbol);
+                    }
+
+                    var attributeTypeSymbols = attributeTypeSymbolsBuilder.ToImmutable();
+                    var streamingContextTypeSymbol = WellKnownTypes.StreamingContext(compilation);
+                    var IDeserializationCallback = WellKnownTypes.IDeserializationCallback(compilation);
+                    // A dictionary from method symbol to set of methods invoked by it directly.
+                    // The bool value in the sub ConcurrentDictionary is not used, use ConcurrentDictionary rather than HashSet just for the concurrency security.
+                    var callGraph = new ConcurrentDictionary<IMethodSymbol, ConcurrentDictionary<IMethodSymbol, bool>>();
+
+                    compilationStartAnalysisContext.RegisterOperationBlockStartAction(
+                        (OperationBlockStartAnalysisContext operationBlockStartAnalysisContext) =>
+                        {
+                            var owningSymbol = operationBlockStartAnalysisContext.OwningSymbol;
+
+                            if (owningSymbol.Kind != SymbolKind.Method)
+                            {
+                                return;
+                            }
+
+                            var methodSymbol = (IMethodSymbol)owningSymbol;
+                            var classSymbol = methodSymbol.ContainingType;
+
+                            if (!classSymbol.GetAttributes().Any(s => s.AttributeClass.Equals(serializableAttributeTypeSymbol)))
+                            {
+                                return;
+                            }
+
+                            var calledMethods = new ConcurrentDictionary<IMethodSymbol, bool>();
+                            callGraph.TryAdd(methodSymbol, calledMethods);
+
+                            operationBlockStartAnalysisContext.RegisterOperationAction(operationContext =>
+                            {
+                                callGraph[methodSymbol].TryAdd((operationContext.Operation as IInvocationOperation).TargetMethod, true);
+                            }, OperationKind.Invocation);
+                        });
+
+                    compilationStartAnalysisContext.RegisterCompilationEndAction(
+                        (CompilationAnalysisContext compilationAnalysisContext) =>
+                        {
+                            var visited = new HashSet<IMethodSymbol>();
+                            var results = new Dictionary<IMethodSymbol, HashSet<IMethodSymbol>>();
+
+                            foreach (var kvp in callGraph)
+                            {
+                                var methodSymbol = kvp.Key;
+                                FindCalledDangerousMethod(methodSymbol, visited, results);
```

Comment:
Why call `FindCalledDangerousMethod` before checking if the method is automatically called?